File: nowcoder/python_version/class05/Problem07_ChooseWork.py
# 为了找到自己满意的工作，牛牛收集了每种工作的难度和报酬。牛牛选工作的标准是在难度不超过自身能力值的情况下，牛牛选择报酬最高的工作。
# class Job {
# public int money;// 该工作的报酬 public int hard; // 该工作的难度
#     public Job(int money, int hard) {
#         this.money = money;
#         this.hard = hard;
# } }
# 给定一个Job类型的数组jobarr，表示所有的工作。给定一个int类型的数组arr，表示所有小伙伴的能力。 返回int类型的数组，表示每一个小伙伴按照牛牛的标准选工作后所能获得的报酬。

# 先将工作按难度排序 去掉 工作难度相同 薪水不是最高的工作 然后 去掉工作难度更大 但薪水还没前一个难度高的工作 最后按这个有序表 排序 查找
import random
class Job:
    def __init__(self, money, hard):
        self.money = money
        self.hard = hard

    def __str__(self):
        return 'hard: ' + str(self.hard)+ ' money: ' + str(self.money)

    def __lt__(self, other):
        return self.hard < other.hard

def getJobs(arr, ability):
    jobs = dict()
    for job in arr:
        if job.hard not in jobs.keys():
            jobs[job.hard] = job.money
        else:
            if jobs[job.hard] < job.money:
                jobs[job.hard] = job.money
    jobs = sorted(jobs.items())
    final = []
    final.append(jobs[0])
    for i in range(1, len(jobs)):
        if jobs[i][1] > final[-1][1]:
            final.append(jobs[i])
    print(final)
    jobs = []
    moneys = []
    for i in final:
        jobs.append(i[0])
        moneys.append(i[1])
    print(jobs, moneys)



if __name__ == '__main__':
    arr = []
    for i in range(10):
        job = Job(random.randrange(1,100), random.randrange(1,5))
        print(job)
        arr.append(job)

    ability = []
    for i in range(20):
        ability.append(random.randrange(1, 5))
    print(ability)
    getJobs(arr, ability)

File: Java/src/classes/class05/Problem07_ChooseWork.java
package classes.class05;

import java.util.*;

public class Problem07_ChooseWork {
    public static class Job {
        public int money;
        public int hard;

        public Job(int money, int hard) {
            this.money = money;
            this.hard = hard;
        }

        @Override
        public String toString() {
            return "("+ this.hard + " " + this.money+")";
        }
    }

    public static class jobComparator implements Comparator<Job>{
        @Override
        public int compare(Job o1, Job o2) {
            return o1.hard != o2.hard ? (o1.hard - o2.hard) : (o2.money - o1.money);
        }
    }

    public static int [] chooseWork(Job [] jobs, int [] ability){
        Arrays.sort(jobs,new jobComparator());

        TreeMap<Integer, Integer> jobMap = new TreeMap<>();
        int maxMoney = 0;
        for (int i = 0; i < jobs.length; i++){
            if(!jobMap.containsKey(jobs[i].hard) && jobs[i].money > maxMoney){
                jobMap.put(jobs[i].hard, jobs[i].money);
                maxMoney = jobs[i].money;
            }
        }


        int [] result = new int[ability.length];
        for(int i = 0; i < ability.length; i++){
            result[i] = jobMap.get(jobMap.floorKey(ability[i]));
        }
        return result;
    }

    public static void printArr(int [] arr){
        for(int i = 0; i < arr.length; i++){
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }


    public static void main(String[] args) {
        Job job1 = new Job(100,1);
        Job job2 = new Job(1000,10);
        Job job3 = new Job(100,10);
        Job job4 = new Job(1001, 1000000000);
        Job job5 = new Job(99, 100000001);
        Job [] jobArr = {job1, job2,job4};
        int [] ability = {9,10,1000000000};
        int [] result = chooseWork(jobArr, ability);
        printArr(result);
    }
}
